### PR TITLE
fix: draft for fixing recursive models in OpenAPI schema generation

### LINKED ISSUE #2429: Bug: OpenAPI schema generation fails with recursive models
### Description

When using recursive models such as the example below, Litestar gets a recursion error when generating an OpenAPI schema. Litestar should support these schemas either by having an implicit recursion limit when generating the docs or through some other means.

### URL to code causing the issue

_No response_

### MCVE

```python
from __future__ import annotations
from litestar import Litestar, get
from pydantic import BaseModel

class A(BaseModel):
    a: A | None = None
    b: B | None = None

class B(BaseModel):
    a: A | None = None
    b: B | None = None

@get("/")
async def test() -> A:
    return A()

app = Litestar([test])
```

For comparison, here is a FastAPI version which works properly:
```py
from __future__ import annotations
from fastapi import FastAPI
from pydantic import BaseModel

class A(BaseModel):
    a: A | None = None
    b: B | None = None

class B(BaseModel):
    a: A | None = None
    b: B | None = None

app = FastAPI()

@app.get("/")
async def test() -> A:
    return A()
```

### Steps to reproduce

```bash
1. Run the litestar API
2. Go to http://127.0.0.1:8000/schema/swagger or a different doc format.
3. See Internal Server Error (or more details if API run with --debug)
4. Refresh, and see an empty schema
```

### Screenshots

_No response_

### Logs

_No response_

### Litestar Version

2.2.0

### Platform

- [ ] Linux
- [ ] Mac
- [X] Windows
- [ ] Other (Please specify in the description above)

### PR DESCRIPTION
**Pull request by jw-can**
### Pull Request Checklist

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [ ] Pre-Commit Checks were ran and passed
- [ ] Tests were ran and passed

### Description

I made a proof of concept that has some additional logic in schema.py, which keeps track of the schemas that are being processed. If a schema B has a property of schema A, which we're currently processing, that property get's ignored to unblock the processing of schema B and break the circularity. The logic keeps track of that ignored property and adds it after the schema A has been processed to restore the circularity.
so if we had A which has a property b which references B and B which has a property a which references A, the flow looks as follows:

1. start processing A
2. process property b -> process B
3. start processing B
4. process property a -> ignore (since we're still processing A)
5. finish processing B
6. finish processing A
7. add previously ignored property a to B

### Close Issue(s)

- Fixes #2429

### CODE REVIEW

File: litestar/contrib/pydantic/pydantic_schema_plugin.py [hunk 1]
```diff
@@ -141,6 +141,48 @@ def __init__(self, prefer_alias: bool = False) -> None:
...
     def is_plugin_supported_type(value: Any) -> bool:
         return isinstance(value, _supported_types) or is_class_and_subclass(value, _supported_types)  # type: ignore
+    
+    def get_field_name(self, field_definition: FieldDefinition, schema_creator: SchemaCreator) -> str:
+        model_fields: dict[str, pydantic.fields.FieldInfo] = {
+            k: getattr(f, "field_info", f)
+            for k, f in getattr(field_definition.annotation, "__fields__", getattr(field_definition.annotation, "model_fields", {})).items()
+        }
+
+        annotation_hints = get_type_hints(field_definition.annotation, include_extras=True)
```
Comment (peterschutt) [inline litestar/contrib/pydantic/pydantic_schema_plugin.py:151]:
```suggestion
        annotation_hints = field_definition.get_type_hints(include_extras=True)
```
`FieldDefinition` has a method for exactly this!

File: litestar/contrib/pydantic/pydantic_schema_plugin.py [hunk 2]
```diff
@@ -141,6 +141,48 @@ def __init__(self, prefer_alias: bool = False) -> None:
...
+            for k, f in model_fields.items()
+        }
+    
+    def get_field_definitions(self, field_definition: FieldDefinition, schema_creator: SchemaCreator) -> list[FieldDefinition]:
+        model_fields: dict[str, pydantic.fields.FieldInfo] = {
+            k: getattr(f, "field_info", f)
+            for k, f in getattr(field_definition.annotation, "__fields__", getattr(field_definition.annotation, "model_fields", {})).items()
+        }
+
+        annotation_hints = get_type_hints(field_definition.annotation, include_extras=True)
```
Comment (peterschutt) [inline litestar/contrib/pydantic/pydantic_schema_plugin.py:171]: Same.

**Review by peterschutt — COMMENTED** [review]
A couple of quick comments. I assume there is still more to push up here?

**Comment by jw-can** [comment]
Hi @peterschutt 
This is a rough draft proposal for a solution to this issue. If the direction is good I'm happy to flesh this out into a proper pull request
(I reached out to __peter__ on discord, is that you?)

**Comment by peterschutt** [comment]
> Hi @peterschutt This is a rough draft proposal for a solution to this issue. If the direction is good I'm happy to flesh this out into a proper pull request (with your suggestions, thanks!)

Thanks, from your summary it sounds like a reasonable approach. Hard to comment without seeing the implementation though.

 (I reached out to **peter** on discord, is that you?)

Yep, that's me!

**Comment by jw-can** [comment]
I'm sorry, I didn't commit the most important file yesterday... was a bit late at night!

**Comment by guacs** [comment]
@jw-can A few points:
- this only seems to cover plugins, but it should cover all the other types we support as well
- adding tests would help in seeing what the expected behavior is which in turn would help in reviewing the code and the approach

**Comment by peterschutt** [comment]
Thanks for  making a start on this! We are closing in favor of #2869 but appreciate your interest and effort!

**Pull request closed without merging**